// frontend/src/game/AudioEngine.js
export class AudioEngine {
  constructor() {
    this.ctx = new (window.AudioContext || window.webkitAudioContext)();
    this.isCleaningUp = false;
    this.cleanupTimer = null;

    // Default values
    this.enabled = true;
    this.sfxVolume = 1.0;
    this.masterVolume = 0.3;
    this.voiceEnabled = true;

    this.masterGain = this.ctx.createGain();
    this.masterGain.connect(this.ctx.destination);

    this.loadSettings();

    this.ambientNodes = [];
    this.musicState = 'menu';
    this.lastAnnouncementAt = 0;
  }

  loadSettings() {
    try {
      const saved = localStorage.getItem('snakey_audio_settings');
      if (saved) {
        const settings = JSON.parse(saved);
        this.enabled = settings.enabled ?? true;
        this.sfxVolume = settings.sfxVolume ?? 1.0;
        this.masterVolume = settings.masterVolume ?? 0.3;
        this.voiceEnabled = settings.voiceEnabled ?? true;
        this.masterGain.gain.value = this.masterVolume;
      } else {
        this.masterGain.gain.value = this.masterVolume;
      }
    } catch (e) {
      console.warn('AudioEngine: Failed to load settings', e);
      this.masterGain.gain.value = this.masterVolume;
    }
  }

  saveSettings() {
    try {
      localStorage.setItem(
        'snakey_audio_settings',
        JSON.stringify({
          enabled: this.enabled,
          sfxVolume: this.sfxVolume,
          masterVolume: this.masterVolume,
          voiceEnabled: this.voiceEnabled,
        })
      );
    } catch (e) {
      console.warn('AudioEngine: Failed to save settings', e);
    }
  }

  resume() {
    if (this.ctx.state === 'suspended') {
      this.ctx.resume();
    }
  }

  setVoiceEnabled(bool) {
    this.voiceEnabled = bool;
    if (!bool && 'speechSynthesis' in window) {
      window.speechSynthesis.cancel();
    }
    this.saveSettings();
  }

  announce(text, options = {}) {
    if (!this.enabled || !this.voiceEnabled || !('speechSynthesis' in window)) return;
    const now = performance.now();
    if (!options.force && now - this.lastAnnouncementAt < 900) return;

    this.lastAnnouncementAt = now;
    if (options.interrupt) {
      window.speechSynthesis.cancel();
    }

    const utterance = new SpeechSynthesisUtterance(text);
    utterance.rate = options.rate ?? 1.03;
    utterance.pitch = options.pitch ?? 0.92;
    utterance.volume = Math.min(1, Math.max(0, this.masterVolume * 0.9));
    window.speechSynthesis.speak(utterance);
  }

  playCinematicSting() {
    if (!this.enabled) return;
    const now = this.ctx.currentTime;
    [220, 330, 440, 660].forEach((freq, index) => {
      const osc = this.ctx.createOscillator();
      const gain = this.ctx.createGain();
      osc.type = index % 2 === 0 ? 'triangle' : 'sine';
      osc.frequency.setValueAtTime(freq, now + index * 0.08);
      gain.gain.setValueAtTime(0, now + index * 0.08);
      gain.gain.linearRampToValueAtTime(0.15 * this.sfxVolume, now + index * 0.08 + 0.04);
      gain.gain.exponentialRampToValueAtTime(0.001, now + index * 0.08 + 0.55);
      osc.connect(gain);
      gain.connect(this.masterGain);
      osc.start(now + index * 0.08);
      osc.stop(now + index * 0.08 + 0.6);
    });
  }

  playEat() {
    if (!this.enabled) return;
    const osc = this.ctx.createOscillator();
    const gain = this.ctx.createGain();

    osc.type = 'sine';
    osc.frequency.setValueAtTime(400, this.ctx.currentTime);
    osc.frequency.exponentialRampToValueAtTime(800, this.ctx.currentTime + 0.1);

    gain.gain.setValueAtTime(0, this.ctx.currentTime);
    gain.gain.linearRampToValueAtTime(0.5 * this.sfxVolume, this.ctx.currentTime + 0.02);
    gain.gain.linearRampToValueAtTime(0, this.ctx.currentTime + 0.1);

    osc.connect(gain);
    gain.connect(this.masterGain);

    osc.start();
    osc.stop(this.ctx.currentTime + 0.1);
  }

  playBoost() {
    if (!this.enabled) return;
    const now = this.ctx.currentTime;
    const osc = this.ctx.createOscillator();
    const overtone = this.ctx.createOscillator();
    const gain = this.ctx.createGain();
    const filter = this.ctx.createBiquadFilter();

    osc.type = 'triangle';
    overtone.type = 'sawtooth';
    osc.frequency.setValueAtTime(170, now);
    osc.frequency.exponentialRampToValueAtTime(110, now + 0.12);
    overtone.frequency.setValueAtTime(320, now);
    overtone.frequency.exponentialRampToValueAtTime(180, now + 0.12);

    filter.type = 'lowpass';
    filter.frequency.setValueAtTime(900, now);

    gain.gain.setValueAtTime(0, now);
    gain.gain.linearRampToValueAtTime(0.16 * this.sfxVolume, now + 0.02);
    gain.gain.exponentialRampToValueAtTime(0.001, now + 0.14);

    osc.connect(filter);
    overtone.connect(filter);
    filter.connect(gain);
    gain.connect(this.masterGain);

    osc.start(now);
    overtone.start(now);
    osc.stop(now + 0.15);
    overtone.stop(now + 0.15);
  }

  playAbility() {
    if (!this.enabled) return;
    const osc = this.ctx.createOscillator();
    const gain = this.ctx.createGain();

    osc.type = 'square';
    osc.frequency.setValueAtTime(600, this.ctx.currentTime);
    osc.frequency.exponentialRampToValueAtTime(1200, this.ctx.currentTime + 0.2);

    gain.gain.setValueAtTime(0, this.ctx.currentTime);
    gain.gain.linearRampToValueAtTime(0.3 * this.sfxVolume, this.ctx.currentTime + 0.05);
    gain.gain.linearRampToValueAtTime(0, this.ctx.currentTime + 0.3);

    osc.connect(gain);
    gain.connect(this.masterGain);

    osc.start();
    osc.stop(this.ctx.currentTime + 0.3);
  }

  playKill(streak = 1) {
    if (!this.enabled) return;
    const now = this.ctx.currentTime;
    const osc = this.ctx.createOscillator();
    const harmony = this.ctx.createOscillator();
    const gain = this.ctx.createGain();

    osc.type = 'sawtooth';
    harmony.type = 'square';
    const baseFreq = Math.min(420, 190 + streak * 28);
    osc.frequency.setValueAtTime(baseFreq, now);
    osc.frequency.exponentialRampToValueAtTime(baseFreq * 0.42, now + 0.38);
    harmony.frequency.setValueAtTime(baseFreq * 1.5, now);
    harmony.frequency.exponentialRampToValueAtTime(baseFreq * 0.7, now + 0.24);

    gain.gain.setValueAtTime(0, now);
    gain.gain.linearRampToValueAtTime(Math.min(0.95, 0.55 + streak * 0.08) * this.sfxVolume, now + 0.03);
    gain.gain.exponentialRampToValueAtTime(0.001, now + 0.5);

    osc.connect(gain);
    harmony.connect(gain);
    gain.connect(this.masterGain);

    osc.start(now);
    harmony.start(now);
    osc.stop(now + 0.5);
    harmony.stop(now + 0.28);

    if (streak === 3) {
      this.announce('Triple elimination.', { interrupt: true, rate: 1.08 });
    } else if (streak >= 5) {
      this.announce(`${streak} kill streak. Arena dominance.`, { interrupt: true, rate: 1.1, force: true });
    }
  }

  playDeath() {
    if (!this.enabled) return;
    const osc = this.ctx.createOscillator();
    const gain = this.ctx.createGain();

    osc.type = 'sawtooth';
    osc.frequency.setValueAtTime(150, this.ctx.currentTime);
    osc.frequency.exponentialRampToValueAtTime(40, this.ctx.currentTime + 0.8);

    gain.gain.setValueAtTime(0, this.ctx.currentTime);
    gain.gain.linearRampToValueAtTime(0.6 * this.sfxVolume, this.ctx.currentTime + 0.1);
    gain.gain.linearRampToValueAtTime(0, this.ctx.currentTime + 0.8);

    osc.connect(gain);
    gain.connect(this.masterGain);

    osc.start();
    osc.stop(this.ctx.currentTime + 0.8);
    this.announce('Snake down.', { interrupt: true, rate: 0.94 });
  }

  playPowerUp(type = 'POWER') {
    if (!this.enabled) return;
    const now = this.ctx.currentTime;
    const frequencyMap = {
      SHIELD: [420, 620, 860],
      MAGNET: [320, 540, 820],
      FRENZY: [460, 720, 1020],
      GHOST: [360, 520, 780],
      POWER: [600, 800, 1200],
    };
    const notes = frequencyMap[type] || frequencyMap.POWER;

    // Arpeggio effect
    notes.forEach((freq, i) => {
      const osc = this.ctx.createOscillator();
      const gain = this.ctx.createGain();
      osc.type = i === notes.length - 1 ? 'triangle' : 'sine';
      osc.frequency.setValueAtTime(freq, now + i * 0.05);
      gain.gain.setValueAtTime(0, now + i * 0.05);
      gain.gain.linearRampToValueAtTime(0.3 * this.sfxVolume, now + i * 0.05 + 0.02);
      gain.gain.linearRampToValueAtTime(0, now + i * 0.05 + 0.2);
      osc.connect(gain);
      gain.connect(this.masterGain);
      osc.start(now + i * 0.05);
      osc.stop(now + i * 0.05 + 0.2);
    });
  }

  createAmbientLayer({ type, frequency, gainValue, detune = 0, filterType = 'lowpass', filterFrequency = 1200 }) {
    const osc = this.ctx.createOscillator();
    const filter = this.ctx.createBiquadFilter();
    const gain = this.ctx.createGain();
    const lfo = this.ctx.createOscillator();
    const lfoGain = this.ctx.createGain();

    osc.type = type;
    osc.frequency.setValueAtTime(frequency, this.ctx.currentTime);
    osc.detune.setValueAtTime(detune, this.ctx.currentTime);

    filter.type = filterType;
    filter.frequency.setValueAtTime(filterFrequency, this.ctx.currentTime);

    gain.gain.setValueAtTime(0, this.ctx.currentTime);
    gain.gain.linearRampToValueAtTime(gainValue, this.ctx.currentTime + 1.2);

    lfo.type = 'sine';
    lfo.frequency.setValueAtTime(0.08 + Math.random() * 0.06, this.ctx.currentTime);
    lfoGain.gain.setValueAtTime(4 + Math.random() * 7, this.ctx.currentTime);
    lfo.connect(lfoGain);
    lfoGain.connect(osc.frequency);

    osc.connect(filter);
    filter.connect(gain);
    gain.connect(this.masterGain);

    osc.start();
    lfo.start();
    return { osc, filter, gain, lfo, lfoGain, baseFrequency: frequency, baseGain: gainValue };
  }

  startAmbient(state = 'menu') {
    this.musicState = state;
    if (!this.enabled) return;
    if (!this.ambientNodes.length) {
      this.ambientNodes = [
        this.createAmbientLayer({ type: 'triangle', frequency: 58, gainValue: 0.035, detune: -3, filterFrequency: 820 }),
        this.createAmbientLayer({ type: 'sine', frequency: 116, gainValue: 0.018, detune: 2, filterFrequency: 1400 }),
        this.createAmbientLayer({ type: 'sawtooth', frequency: 232, gainValue: 0.008, detune: 5, filterFrequency: 900 }),
      ];
    }

    this.applyMusicState(state);
  }

  applyMusicState(state) {
    if (!this.ambientNodes.length) {
      return;
    }

    const profiles = {
      menu: [
        { frequency: 58, gain: 0.035 },
        { frequency: 116, gain: 0.02 },
        { frequency: 232, gain: 0.008 },
      ],
      briefing: [
        { frequency: 64, gain: 0.04 },
        { frequency: 128, gain: 0.024 },
        { frequency: 256, gain: 0.01 },
      ],
      combat: [
        { frequency: 72, gain: 0.05 },
        { frequency: 144, gain: 0.028 },
        { frequency: 288, gain: 0.014 },
      ],
      results: [
        { frequency: 52, gain: 0.028 },
        { frequency: 104, gain: 0.018 },
        { frequency: 208, gain: 0.006 },
      ],
    };

    const profile = profiles[state] || profiles.menu;
    this.ambientNodes.forEach((node, index) => {
      const next = profile[index] || profile[profile.length - 1];
      node.osc.frequency.setTargetAtTime(next.frequency, this.ctx.currentTime, 0.8);
      node.gain.gain.setTargetAtTime(next.gain, this.ctx.currentTime, 0.8);
    });
  }

  setMusicState(state) {
    this.musicState = state;
    this.startAmbient(state);
  }

  stopAmbient() {
    if (this.ambientNodes.length) {
      this.ambientNodes.forEach((node) => {
        node.gain.gain.linearRampToValueAtTime(0, this.ctx.currentTime + 1);
      });
      setTimeout(() => {
        this.ambientNodes.forEach((node) => {
          node.osc.stop();
          node.lfo.stop();
        });
        this.ambientNodes = [];
      }, 1100);
    }
  }

  setMasterVolume(v) {
    this.masterVolume = v;
    this.masterGain.gain.setTargetAtTime(v, this.ctx.currentTime, 0.1);
    this.saveSettings();
  }

  setSfxVolume(v) {
    this.sfxVolume = v;
    this.saveSettings();
  }

  setEnabled(bool) {
    this.enabled = bool;
    if (!bool) this.stopAmbient();
    else if (this.ctx.state === 'running') this.startAmbient(this.musicState);
    this.saveSettings();
  }

  cleanup() {
    if (this.isCleaningUp || !this.ctx) return;
    this.isCleaningUp = true;

    this.stopAmbient();

    if (this.cleanupTimer) {
      clearTimeout(this.cleanupTimer);
      this.cleanupTimer = null;
    }

    const ctx = this.ctx;
    this.cleanupTimer = setTimeout(() => {
      if (ctx && ctx.state !== 'closed') {
        const closePromise = ctx.close();
        if (closePromise && typeof closePromise.catch === 'function') {
          closePromise.catch((err) => {
            if (!err || err.name !== 'InvalidStateError') {
              console.warn('AudioEngine: Failed to close AudioContext', err);
            }
          });
        }
      }
      this.cleanupTimer = null;
    }, 1200);
  }
}
