// frontend/src/game/ProfileManager.js
export const CHALLENGES = [
  {
    id: 'shadow_stalker',
    name: 'Shadow Stalker',
    description: 'Score 4 kills in a single match.',
    target: 4,
    metric: 'matchKills',
    progressType: 'best',
  },
  {
    id: 'skybreaker',
    name: 'Skybreaker',
    description: 'Reach 1,500 mass in one run.',
    target: 1500,
    metric: 'maxMass',
    progressType: 'best',
  },
  {
    id: 'star_forge',
    name: 'Star Forge',
    description: 'Accumulate 40 total kills.',
    target: 40,
    metric: 'totalKills',
    progressType: 'total',
  },
  {
    id: 'coil_marathon',
    name: 'Coil Marathon',
    description: 'Play 15 matches.',
    target: 15,
    metric: 'totalMatches',
    progressType: 'total',
  },
  {
    id: 'royal_audience',
    name: 'Royal Audience',
    description: 'Claim 8 daily rewards.',
    target: 8,
    metric: 'dailyClaims',
    progressType: 'total',
  },
];

export const DAILY_REWARDS = [
  { day: 1, xp: 300, label: 'Warm-up cache' },
  { day: 2, xp: 500, label: 'Neon ration pack' },
  { day: 3, xp: 750, label: 'Smuggler stash' },
  { day: 4, xp: 1050, label: 'Heroic supply drop' },
  { day: 5, xp: 1400, label: 'Royal sigil crate' },
  { day: 6, xp: 1800, label: 'Galactic armory' },
  { day: 7, xp: 2400, label: 'Mythic vault' },
];

export const SKINS = [
  {
    id: 'default',
    name: 'Volt Classic',
    line: 'Arcade Core',
    tagline: 'Clean neon lanes with arcade energy.',
    rarity: 'Core',
    requiredLevel: 1,
    unlock: { type: 'level', level: 1 },
    baseColor: '#00ffcc',
    secondaryColor: '#0ea5e9',
    accentColor: '#d9fffb',
    trailColor: '#7efcf2',
    pattern: 'glow',
    trailStyle: 'glow',
    deathEffect: 'pulse',
  },
  {
    id: 'neon_pink',
    name: 'Neon Bloom',
    line: 'Arcade Core',
    tagline: 'Hot-pink pulse with club-floor shine.',
    rarity: 'Rare',
    requiredLevel: 2,
    unlock: { type: 'level', level: 2 },
    baseColor: '#ff4fd8',
    secondaryColor: '#ff99f2',
    accentColor: '#fff0fb',
    trailColor: '#ffb4ef',
    pattern: 'neon',
    trailStyle: 'sparkle',
    deathEffect: 'petal',
  },
  {
    id: 'ember_stripe',
    name: 'Ember Stripe',
    line: 'Arcade Core',
    tagline: 'A predator pattern made for kill streaks.',
    rarity: 'Rare',
    requiredLevel: 4,
    unlock: { type: 'level', level: 4 },
    baseColor: '#ff8a00',
    secondaryColor: '#2f1400',
    accentColor: '#ffd37a',
    trailColor: '#ffbe63',
    pattern: 'stripe',
    trailStyle: 'ember',
    deathEffect: 'flare',
  },
  {
    id: 'metro_sentinel',
    name: 'Metro Sentinel',
    line: 'Comic Hero',
    tagline: 'Skyline steel, cape-energy glow, big protector aura.',
    rarity: 'Epic',
    requiredLevel: 6,
    unlock: { type: 'level', level: 6 },
    baseColor: '#3b82f6',
    secondaryColor: '#b91c1c',
    accentColor: '#fef3c7',
    trailColor: '#93c5fd',
    pattern: 'royal',
    trailStyle: 'heroic',
    deathEffect: 'shockwave',
  },
  {
    id: 'panda_puff',
    name: 'Panda Puff',
    line: 'Animal Parade',
    tagline: 'Creamy panda patches with a bedtime-story bounce.',
    rarity: 'Rare',
    requiredLevel: 3,
    unlock: { type: 'level', level: 3 },
    baseColor: '#f8fafc',
    secondaryColor: '#1f2937',
    accentColor: '#fbcfe8',
    trailColor: '#fde68a',
    pattern: 'ethereal',
    trailStyle: 'sparkle',
    deathEffect: 'confetti',
  },
  {
    id: 'bunny_blossom',
    name: 'Bunny Blossom',
    line: 'Animal Parade',
    tagline: 'Soft sakura pink with cheerful hop-energy stripes.',
    rarity: 'Rare',
    requiredLevel: 5,
    unlock: { type: 'level', level: 5 },
    baseColor: '#f9a8d4',
    secondaryColor: '#f472b6',
    accentColor: '#fff1f2',
    trailColor: '#fecdd3',
    pattern: 'stripe',
    trailStyle: 'sparkle',
    deathEffect: 'petal',
  },
  {
    id: 'phantom_silk',
    name: 'Phantom Silk',
    line: 'Shadow Operative',
    tagline: 'Icy-white shimmer with spectral fade.',
    rarity: 'Epic',
    requiredLevel: 7,
    unlock: { type: 'level', level: 7 },
    baseColor: '#f8fafc',
    secondaryColor: '#bcd4ff',
    accentColor: '#ffffff',
    trailColor: '#dce9ff',
    pattern: 'ethereal',
    trailStyle: 'mist',
    deathEffect: 'shatter',
  },
  {
    id: 'otter_pop',
    name: 'Otter Pop',
    line: 'Animal Parade',
    tagline: 'Caramel river tones with a playful bubble trail.',
    rarity: 'Epic',
    requiredLevel: 8,
    unlock: { type: 'level', level: 8 },
    baseColor: '#c08457',
    secondaryColor: '#7c3f00',
    accentColor: '#fde68a',
    trailColor: '#fdba74',
    pattern: 'glow',
    trailStyle: 'comet',
    deathEffect: 'flare',
  },
  {
    id: 'midnight_vigil',
    name: 'Midnight Vigil',
    line: 'Comic Hero',
    tagline: 'Gothic black armor with rooftop hunter energy.',
    rarity: 'Legendary',
    requiredLevel: 0,
    unlock: { type: 'challenge', challengeId: 'shadow_stalker' },
    baseColor: '#0f172a',
    secondaryColor: '#312e81',
    accentColor: '#93c5fd',
    trailColor: '#818cf8',
    pattern: 'cosmic',
    trailStyle: 'shadow',
    deathEffect: 'batburst',
  },
  {
    id: 'corgi_curl',
    name: 'Corgi Curl',
    line: 'Animal Parade',
    tagline: 'Golden loaf colors with a tiny royal zoomie streak.',
    rarity: 'Epic',
    requiredLevel: 10,
    unlock: { type: 'dailyClaims', claims: 3 },
    baseColor: '#f59e0b',
    secondaryColor: '#92400e',
    accentColor: '#fff7ed',
    trailColor: '#fdba74',
    pattern: 'royal',
    trailStyle: 'sunfire',
    deathEffect: 'crownflare',
  },
  {
    id: 'sunforge',
    name: 'Sunforge',
    line: 'Regal Warband',
    tagline: 'Liquid gold plating with trophy glow.',
    rarity: 'Legendary',
    requiredLevel: 11,
    unlock: { type: 'level', level: 11 },
    baseColor: '#ffd54a',
    secondaryColor: '#ff9f1c',
    accentColor: '#fff4bf',
    trailColor: '#ffe08a',
    pattern: 'royal',
    trailStyle: 'sunfire',
    deathEffect: 'crownflare',
  },
  {
    id: 'kitty_cloud',
    name: 'Kitty Cloud',
    line: 'Animal Parade',
    tagline: 'Lavender tabby glimmer with sleepy-star highlights.',
    rarity: 'Legendary',
    requiredLevel: 0,
    unlock: { type: 'challenge', challengeId: 'coil_marathon' },
    baseColor: '#c4b5fd',
    secondaryColor: '#7c3aed',
    accentColor: '#f8fafc',
    trailColor: '#ddd6fe',
    pattern: 'cosmic',
    trailStyle: 'stardust',
    deathEffect: 'glassbloom',
  },
  {
    id: 'foxglove_fizz',
    name: 'Foxglove Fizz',
    line: 'Animal Parade',
    tagline: 'Amber fox tones with bright woodland spark.',
    rarity: 'Epic',
    requiredLevel: 0,
    unlock: { type: 'dailyStreak', days: 4 },
    baseColor: '#fb923c',
    secondaryColor: '#9a3412',
    accentColor: '#ffedd5',
    trailColor: '#fdba74',
    pattern: 'neon',
    trailStyle: 'ember',
    deathEffect: 'flare',
  },
  {
    id: 'dawn_saber',
    name: 'Dawn Saber',
    line: 'Galactic Knight',
    tagline: 'Monastic plasma glow with disciplined blade lines.',
    rarity: 'Epic',
    requiredLevel: 12,
    unlock: { type: 'level', level: 12 },
    baseColor: '#38bdf8',
    secondaryColor: '#1d4ed8',
    accentColor: '#eff6ff',
    trailColor: '#7dd3fc',
    pattern: 'frost',
    trailStyle: 'lightslice',
    deathEffect: 'nova',
  },
  {
    id: 'storyglow',
    name: 'Storyglow',
    line: 'Enchanted Kingdom',
    tagline: 'Fairy-book pastels with a stage-light finish.',
    rarity: 'Rare',
    requiredLevel: 9,
    unlock: { type: 'level', level: 9 },
    baseColor: '#f9a8d4',
    secondaryColor: '#a78bfa',
    accentColor: '#fef3c7',
    trailColor: '#fbcfe8',
    pattern: 'neon',
    trailStyle: 'sparkle',
    deathEffect: 'confetti',
  },
  {
    id: 'smuggler_flux',
    name: 'Smuggler Flux',
    line: 'Space Rogue',
    tagline: 'Quick-draw chrome, rust-red trim, no questions asked.',
    rarity: 'Epic',
    requiredLevel: 0,
    unlock: { type: 'dailyStreak', days: 3 },
    baseColor: '#f97316',
    secondaryColor: '#334155',
    accentColor: '#fde68a',
    trailColor: '#fdba74',
    pattern: 'stripe',
    trailStyle: 'comet',
    deathEffect: 'scrapburst',
  },
  {
    id: 'aurora_coil',
    name: 'Aurora Coil',
    line: 'Polar Myth',
    tagline: 'Frozen cyan washed in northern light.',
    rarity: 'Epic',
    requiredLevel: 14,
    unlock: { type: 'level', level: 14 },
    baseColor: '#73f0ff',
    secondaryColor: '#7c5cff',
    accentColor: '#e8ffff',
    trailColor: '#b0fbff',
    pattern: 'frost',
    trailStyle: 'aurora',
    deathEffect: 'shatter',
  },
  {
    id: 'glass_aurora',
    name: 'Glass Aurora',
    line: 'Enchanted Princess',
    tagline: 'Crystal palace colorwork with graceful frost trails.',
    rarity: 'Legendary',
    requiredLevel: 0,
    unlock: { type: 'dailyStreak', days: 5 },
    baseColor: '#c084fc',
    secondaryColor: '#67e8f9',
    accentColor: '#fef9c3',
    trailColor: '#ddd6fe',
    pattern: 'ethereal',
    trailStyle: 'stardust',
    deathEffect: 'glassbloom',
  },
  {
    id: 'venom_circuit',
    name: 'Venom Circuit',
    line: 'Biohazard Vanguard',
    tagline: 'Toxic green scales with biotech veins.',
    rarity: 'Epic',
    requiredLevel: 17,
    unlock: { type: 'level', level: 17 },
    baseColor: '#7dff5a',
    secondaryColor: '#134e1d',
    accentColor: '#dfffd2',
    trailColor: '#bbf7d0',
    pattern: 'venom',
    trailStyle: 'toxic',
    deathEffect: 'acid',
  },
  {
    id: 'solar_vanguard',
    name: 'Solar Vanguard',
    line: 'Comic Hero',
    tagline: 'Iconic champion energy with star-lit armor seams.',
    rarity: 'Mythic',
    requiredLevel: 0,
    unlock: { type: 'challenge', challengeId: 'skybreaker' },
    baseColor: '#f43f5e',
    secondaryColor: '#1d4ed8',
    accentColor: '#fef08a',
    trailColor: '#fca5a5',
    pattern: 'royal',
    trailStyle: 'heroic',
    deathEffect: 'supernova',
  },
  {
    id: 'cosmic_drifter',
    name: 'Nebula Drifter',
    line: 'Space Rogue',
    tagline: 'Deep-space blues dusted with stars.',
    rarity: 'Epic',
    requiredLevel: 20,
    unlock: { type: 'level', level: 20 },
    baseColor: '#5b6cff',
    secondaryColor: '#111827',
    accentColor: '#d8dfff',
    trailColor: '#93c5fd',
    pattern: 'cosmic',
    trailStyle: 'stardust',
    deathEffect: 'warp',
  },
  {
    id: 'void_templar',
    name: 'Void Templar',
    line: 'Galactic Knight',
    tagline: 'Ancient order robes wrapped in dark-violet plasma.',
    rarity: 'Mythic',
    requiredLevel: 0,
    unlock: { type: 'challenge', challengeId: 'star_forge' },
    baseColor: '#312e81',
    secondaryColor: '#111827',
    accentColor: '#c4b5fd',
    trailColor: '#a5b4fc',
    pattern: 'prism',
    trailStyle: 'lightslice',
    deathEffect: 'rift',
  },
  {
    id: 'riftburn',
    name: 'Riftburn',
    line: 'Apocalypse Forge',
    tagline: 'Cracked lava armor with a molten core.',
    rarity: 'Legendary',
    requiredLevel: 23,
    unlock: { type: 'level', level: 23 },
    baseColor: '#ff5b2e',
    secondaryColor: '#4a1200',
    accentColor: '#ffd1b8',
    trailColor: '#fdba74',
    pattern: 'magma',
    trailStyle: 'ember',
    deathEffect: 'magma',
  },
  {
    id: 'crownspell',
    name: 'Crownspell',
    line: 'Enchanted Princess',
    tagline: 'Royal midnight silk with spellbound gold trim.',
    rarity: 'Mythic',
    requiredLevel: 0,
    unlock: { type: 'challenge', challengeId: 'royal_audience' },
    baseColor: '#6d28d9',
    secondaryColor: '#1e1b4b',
    accentColor: '#fcd34d',
    trailColor: '#ddd6fe',
    pattern: 'royal',
    trailStyle: 'sparkle',
    deathEffect: 'crownflare',
  },
  {
    id: 'dreamkeep',
    name: 'Dreamkeep',
    line: 'Enchanted Kingdom',
    tagline: 'Castle-lights fantasy with a polished storybook finish.',
    rarity: 'Mythic',
    requiredLevel: 0,
    unlock: { type: 'dailyStreak', days: 7 },
    baseColor: '#ec4899',
    secondaryColor: '#6366f1',
    accentColor: '#fde68a',
    trailColor: '#f9a8d4',
    pattern: 'prism',
    trailStyle: 'storybook',
    deathEffect: 'confetti',
  },
  {
    id: 'outrunner_zero',
    name: 'Outrunner Zero',
    line: 'Space Rogue',
    tagline: 'Hyperspeed white hull with contraband neon wake.',
    rarity: 'Mythic',
    requiredLevel: 0,
    unlock: { type: 'challenge', challengeId: 'coil_marathon' },
    baseColor: '#e2e8f0',
    secondaryColor: '#0f172a',
    accentColor: '#22d3ee',
    trailColor: '#67e8f9',
    pattern: 'prism',
    trailStyle: 'comet',
    deathEffect: 'warp',
  },
];

const CHALLENGE_LOOKUP = Object.fromEntries(CHALLENGES.map((challenge) => [challenge.id, challenge]));

const SKIN_LOOKUP = Object.fromEntries(SKINS.map((skin) => [skin.id, skin]));

function getDateKey(date = new Date()) {
  const year = date.getFullYear();
  const month = `${date.getMonth() + 1}`.padStart(2, '0');
  const day = `${date.getDate()}`.padStart(2, '0');
  return `${year}-${month}-${day}`;
}

function diffCalendarDays(previousDateKey, nextDateKey) {
  if (!previousDateKey || !nextDateKey) return null;
  const previousDate = new Date(`${previousDateKey}T00:00:00`);
  const nextDate = new Date(`${nextDateKey}T00:00:00`);
  const diffMs = nextDate.getTime() - previousDate.getTime();
  return Math.round(diffMs / 86400000);
}

function createDefaultProfile() {
  return {
    name: 'Player',
    xp: 0,
    level: 1,
    performancePreset: 'adaptive',
    selectedSkin: 'default',
    unlockedSkins: ['default'],
    completedChallenges: [],
    challengeProgress: {},
    stats: {
      totalMatches: 0,
      totalKills: 0,
      bestMass: 50,
      bestXp: 0,
    },
    dailyRewards: {
      lastClaimDate: null,
      streak: 0,
      bestStreak: 0,
      totalClaims: 0,
    },
  };
}

function getMetricValue(profile, matchSummary, metric) {
  switch (metric) {
    case 'matchKills':
      return matchSummary.matchKills || 0;
    case 'maxMass':
      return matchSummary.maxMass || 0;
    case 'totalKills':
      return profile.stats.totalKills || 0;
    case 'totalMatches':
      return profile.stats.totalMatches || 0;
    case 'dailyClaims':
      return profile.dailyRewards.totalClaims || 0;
    default:
      return 0;
  }
}

function ensureProfileShape(profile) {
  const defaultProfile = createDefaultProfile();
  const mergedProfile = {
    ...defaultProfile,
    ...profile,
    stats: {
      ...defaultProfile.stats,
      ...(profile.stats || {}),
    },
    dailyRewards: {
      ...defaultProfile.dailyRewards,
      ...(profile.dailyRewards || {}),
    },
  };

  mergedProfile.completedChallenges = Array.from(new Set(mergedProfile.completedChallenges || []));
  mergedProfile.challengeProgress = { ...(mergedProfile.challengeProgress || {}) };
  mergedProfile.level = ProfileManager.calculateLevel(mergedProfile.xp);
  mergedProfile.performancePreset = mergedProfile.performancePreset === 'ultra' ? 'ultra' : 'adaptive';

  mergedProfile.unlockedSkins = SKINS.filter((skin) => isSkinUnlocked(skin, mergedProfile)).map(
    (skin) => skin.id
  );

  if (!mergedProfile.unlockedSkins.includes(mergedProfile.selectedSkin)) {
    mergedProfile.selectedSkin = 'default';
  }

  return mergedProfile;
}

function isSkinUnlocked(skin, profile) {
  const unlock = skin.unlock || { type: 'level', level: skin.requiredLevel || 1 };
  switch (unlock.type) {
    case 'challenge':
      return (profile.completedChallenges || []).includes(unlock.challengeId);
    case 'dailyStreak':
      return (profile.dailyRewards?.bestStreak || 0) >= unlock.days;
    case 'dailyClaims':
      return (profile.dailyRewards?.totalClaims || 0) >= unlock.claims;
    case 'level':
    default:
      return (profile.level || 1) >= (unlock.level || skin.requiredLevel || 1);
  }
}

function updateChallengeState(profile, matchSummary = {}) {
  const newlyCompleted = [];
  for (const challenge of CHALLENGES) {
    const currentMetricValue = getMetricValue(profile, matchSummary, challenge.metric);
    const previousProgress = profile.challengeProgress[challenge.id] || 0;
    const nextProgress =
      challenge.progressType === 'best'
        ? Math.max(previousProgress, currentMetricValue)
        : currentMetricValue;

    profile.challengeProgress[challenge.id] = nextProgress;

    if (nextProgress >= challenge.target && !profile.completedChallenges.includes(challenge.id)) {
      profile.completedChallenges.push(challenge.id);
      newlyCompleted.push(challenge.id);
    }
  }

  return newlyCompleted;
}

export function getSkinById(id) {
  return SKIN_LOOKUP[id] || SKIN_LOOKUP.default;
}

export function getChallengeById(id) {
  return CHALLENGE_LOOKUP[id] || null;
}

export function getSkinUnlockRequirementText(skin) {
  const unlock = skin.unlock || { type: 'level', level: skin.requiredLevel || 1 };
  switch (unlock.type) {
    case 'challenge': {
      const challenge = getChallengeById(unlock.challengeId);
      return challenge ? challenge.description : 'Complete a hidden challenge';
    }
    case 'dailyStreak':
      return `Claim ${unlock.days} daily rewards in a row`;
    case 'dailyClaims':
      return `Claim ${unlock.claims} total daily rewards`;
    case 'level':
    default:
      return `Reach Lv ${unlock.level || skin.requiredLevel || 1}`;
  }
}

export function getDailyRewardStatus(profile, date = new Date()) {
  const safeProfile = ensureProfileShape(profile || createDefaultProfile());
  const todayKey = getDateKey(date);
  const lastClaimDate = safeProfile.dailyRewards.lastClaimDate;
  const claimedToday = lastClaimDate === todayKey;
  const previousDiff = diffCalendarDays(lastClaimDate, todayKey);
  const nextStreak = claimedToday
    ? safeProfile.dailyRewards.streak || 0
    : previousDiff === 1
    ? (safeProfile.dailyRewards.streak || 0) + 1
    : 1;
  const reward = DAILY_REWARDS[Math.min(nextStreak, DAILY_REWARDS.length) - 1];

  return {
    canClaim: !claimedToday,
    claimedToday,
    currentStreak: safeProfile.dailyRewards.streak || 0,
    nextStreak,
    reward,
  };
}

export class ProfileManager {
  static getProfile() {
    console.log('ProfileManager: getProfile called');
    const defaultProfile = createDefaultProfile();

    // localStorage may not exist in some environments (SSR, tests, incognito)
    if (typeof localStorage === 'undefined' || localStorage === null) {
      console.warn('ProfileManager: localStorage unavailable, using default profile');
      return defaultProfile;
    }

    try {
      const stored = localStorage.getItem('snakey_profile');
      if (stored) {
        const parsed = JSON.parse(stored);
        // guard in case JSON.parse returns null/undefined
        if (parsed && typeof parsed === 'object') {
          return ensureProfileShape({ ...defaultProfile, ...parsed });
        }
      }
    } catch (e) {
      console.error('Error reading profile', e);
      try {
        localStorage.removeItem('snakey_profile');
      } catch {
        return defaultProfile;
      }
    }
    return defaultProfile;
  }

  static saveProfile(profile) {
    localStorage.setItem('snakey_profile', JSON.stringify(ensureProfileShape(profile)));
  }

  static calculateLevel(xp) {
    // Basic curve: 1000 XP per level
    return Math.floor(Math.sqrt(xp / 500)) + 1;
  }

  static getLevelProgress(xp) {
    const currentLevel = this.calculateLevel(xp);
    const xpForCurrent = 500 * Math.pow(currentLevel - 1, 2);
    const xpForNext = 500 * Math.pow(currentLevel, 2);

    return {
      currentLevel,
      xpIntoLevel: xp - xpForCurrent,
      xpRequired: xpForNext - xpForCurrent,
      percent: ((xp - xpForCurrent) / (xpForNext - xpForCurrent)) * 100,
    };
  }

  static addXP(amount, matchSummary = {}) {
    const p = ensureProfileShape(this.getProfile());
    const oldLevel = p.level;
    const previousUnlocks = new Set(p.unlockedSkins || []);

    p.xp += amount;
    p.level = this.calculateLevel(p.xp);
    p.stats.totalMatches += 1;
    p.stats.totalKills += matchSummary.matchKills || 0;
    p.stats.bestMass = Math.max(p.stats.bestMass || 50, matchSummary.maxMass || 0);
    p.stats.bestXp = Math.max(p.stats.bestXp || 0, amount || 0);

    const newChallenges = updateChallengeState(p, matchSummary);
    const normalizedProfile = ensureProfileShape(p);
    const unlockedThisMatch = normalizedProfile.unlockedSkins.filter(
      (skinId) => !previousUnlocks.has(skinId)
    );

    this.saveProfile(normalizedProfile);

    return {
      leveledUp: normalizedProfile.level > oldLevel,
      newLevel: normalizedProfile.level,
      newUnlocks: unlockedThisMatch,
      newChallenges,
      profile: normalizedProfile,
    };
  }

  static claimDailyReward(date = new Date()) {
    const profile = ensureProfileShape(this.getProfile());
    const status = getDailyRewardStatus(profile, date);

    if (!status.canClaim) {
      return {
        claimed: false,
        reason: 'already-claimed',
        reward: status.reward,
        profile,
        newUnlocks: [],
      };
    }

    const previousUnlocks = new Set(profile.unlockedSkins || []);
    const todayKey = getDateKey(date);
    profile.dailyRewards.lastClaimDate = todayKey;
    profile.dailyRewards.streak = status.nextStreak;
    profile.dailyRewards.bestStreak = Math.max(
      profile.dailyRewards.bestStreak || 0,
      status.nextStreak
    );
    profile.dailyRewards.totalClaims = (profile.dailyRewards.totalClaims || 0) + 1;
    profile.xp += status.reward.xp;
    profile.level = this.calculateLevel(profile.xp);

    updateChallengeState(profile, {});
    const normalizedProfile = ensureProfileShape(profile);
    const newUnlocks = normalizedProfile.unlockedSkins.filter((skinId) => !previousUnlocks.has(skinId));

    this.saveProfile(normalizedProfile);

    return {
      claimed: true,
      reward: status.reward,
      profile: normalizedProfile,
      newUnlocks,
      newLevel: normalizedProfile.level,
    };
  }

  static grantBonusXP(amount, source = 'bonus') {
    const profile = ensureProfileShape(this.getProfile());
    const previousUnlocks = new Set(profile.unlockedSkins || []);
    const oldLevel = profile.level;

    profile.xp += amount;
    profile.level = this.calculateLevel(profile.xp);
    profile.stats.bestXp = Math.max(profile.stats.bestXp || 0, amount || 0);

    const normalizedProfile = ensureProfileShape(profile);
    const newUnlocks = normalizedProfile.unlockedSkins.filter((skinId) => !previousUnlocks.has(skinId));
    this.saveProfile(normalizedProfile);

    return {
      source,
      leveledUp: normalizedProfile.level > oldLevel,
      newLevel: normalizedProfile.level,
      newUnlocks,
      profile: normalizedProfile,
      xpGranted: amount,
    };
  }

  /**
   * Completely wipes the stored profile and returns the default object.
   * Useful for debugging or recovering from corrupted storage.
   */
  static resetProfile() {
    try {
      if (typeof localStorage !== 'undefined' && localStorage !== null) {
        localStorage.removeItem('snakey_profile');
      }
    } catch (e) {
      console.warn('ProfileManager.resetProfile: could not clear localStorage', e);
    }
    return this.getProfile();
  }
}
