// frontend/src/game/Renderer.js
import { getSkinById } from './ProfileManager'
import { getThemePreset } from './ModeThemes'

export class Renderer {
  constructor(ctx) {
    this.ctx = ctx;
    this.isMobile =
      /Android|iPhone|iPad|iPod/i.test(navigator.userAgent) || 'ontouchstart' in window;
    this.lowEffects = this.isMobile || window.innerWidth < 1000;
  }

  hexToRgba(hex, alpha) {
    if (typeof hex !== 'string' || !hex.startsWith('#') || hex.length !== 7) {
      return `rgba(255, 255, 255, ${alpha})`
    }

    const value = hex.slice(1)
    const red = Number.parseInt(value.slice(0, 2), 16)
    const green = Number.parseInt(value.slice(2, 4), 16)
    const blue = Number.parseInt(value.slice(4, 6), 16)
    return `rgba(${red}, ${green}, ${blue}, ${alpha})`
  }

  roundRect(x, y, width, height, radius) {
    this.ctx.beginPath()
    this.ctx.moveTo(x + radius, y)
    this.ctx.lineTo(x + width - radius, y)
    this.ctx.quadraticCurveTo(x + width, y, x + width, y + radius)
    this.ctx.lineTo(x + width, y + height - radius)
    this.ctx.quadraticCurveTo(x + width, y + height, x + width - radius, y + height)
    this.ctx.lineTo(x + radius, y + height)
    this.ctx.quadraticCurveTo(x, y + height, x, y + height - radius)
    this.ctx.lineTo(x, y + radius)
    this.ctx.quadraticCurveTo(x, y, x + radius, y)
    this.ctx.closePath()
  }

  getSkinStyle(skinId, fallbackColor) {
    const skin = getSkinById(skinId)
    return {
      ...skin,
      baseColor: skin.baseColor || fallbackColor || '#00ffcc',
      secondaryColor: skin.secondaryColor || fallbackColor || '#0ea5e9',
      accentColor: skin.accentColor || '#ffffff',
      pattern: skin.pattern || 'glow',
      trailColor: skin.trailColor || skin.baseColor || fallbackColor || '#00ffcc',
      trailStyle: skin.trailStyle || 'glow',
      deathEffect: skin.deathEffect || 'pulse',
    }
  }

  createSkinFill(x, y, radius, skin, time, isHead = false) {
    if (skin.pattern === 'prism') {
      const hue = Math.floor((time / 18) % 360)
      const gradient = this.ctx.createLinearGradient(x - radius, y - radius, x + radius, y + radius)
      gradient.addColorStop(0, `hsl(${hue}, 90%, 62%)`)
      gradient.addColorStop(0.4, `hsl(${(hue + 70) % 360}, 92%, 64%)`)
      gradient.addColorStop(1, `hsl(${(hue + 150) % 360}, 88%, 58%)`)
      return gradient
    }

    const gradient = this.ctx.createRadialGradient(
      x - radius * 0.45,
      y - radius * 0.55,
      Math.max(2, radius * 0.15),
      x,
      y,
      radius * (isHead ? 1.1 : 1)
    )

    gradient.addColorStop(0, skin.accentColor)
    gradient.addColorStop(0.45, skin.baseColor)
    gradient.addColorStop(1, skin.secondaryColor)
    return gradient
  }

  drawSkinOverlay(x, y, radius, skin, angle, time, intensity = 1) {
    this.ctx.save()
    this.ctx.beginPath()
    this.ctx.arc(x, y, radius, 0, Math.PI * 2)
    this.ctx.clip()
    this.ctx.globalAlpha = 0.65 * intensity

    switch (skin.pattern) {
      case 'stripe': {
        this.ctx.strokeStyle = this.hexToRgba(skin.secondaryColor, 0.85)
        this.ctx.lineWidth = Math.max(2, radius * 0.22)
        for (let offset = -radius * 1.1; offset <= radius * 1.1; offset += radius * 0.65) {
          this.ctx.beginPath()
          this.ctx.moveTo(x - radius * 1.2, y + offset)
          this.ctx.lineTo(x + radius * 1.2, y + offset - radius * 0.9)
          this.ctx.stroke()
        }
        break
      }
      case 'ethereal': {
        this.ctx.fillStyle = this.hexToRgba('#ffffff', 0.18)
        this.ctx.beginPath()
        this.ctx.arc(x - radius * 0.2, y - radius * 0.18, radius * 0.75, 0, Math.PI * 2)
        this.ctx.fill()
        this.ctx.strokeStyle = this.hexToRgba(skin.secondaryColor, 0.4)
        this.ctx.lineWidth = Math.max(1.5, radius * 0.12)
        this.ctx.beginPath()
        this.ctx.arc(x, y, radius * 0.82, angle - 1.6, angle + 1.4)
        this.ctx.stroke()
        break
      }
      case 'frost': {
        this.ctx.strokeStyle = this.hexToRgba('#ffffff', 0.72)
        this.ctx.lineWidth = Math.max(1.5, radius * 0.11)
        for (let i = -1; i <= 1; i++) {
          this.ctx.beginPath()
          this.ctx.moveTo(x - radius * 0.75, y + i * radius * 0.18)
          this.ctx.lineTo(x + radius * 0.1, y - i * radius * 0.08)
          this.ctx.lineTo(x + radius * 0.65, y + i * radius * 0.22)
          this.ctx.stroke()
        }
        break
      }
      case 'venom': {
        this.ctx.fillStyle = this.hexToRgba(skin.secondaryColor, 0.72)
        ;[
          [-0.35, -0.18, 0.22],
          [0.18, 0.22, 0.16],
          [0.35, -0.28, 0.14],
        ].forEach(([dx, dy, scale]) => {
          this.ctx.beginPath()
          this.ctx.arc(x + radius * dx, y + radius * dy, radius * scale, 0, Math.PI * 2)
          this.ctx.fill()
        })
        break
      }
      case 'cosmic': {
        this.ctx.fillStyle = this.hexToRgba('#ffffff', 0.9)
        ;[
          [-0.3, -0.22, 0.08],
          [0.18, 0.12, 0.06],
          [0.34, -0.05, 0.05],
        ].forEach(([dx, dy, scale]) => {
          this.ctx.beginPath()
          this.ctx.arc(x + radius * dx, y + radius * dy, radius * scale, 0, Math.PI * 2)
          this.ctx.fill()
        })
        this.ctx.strokeStyle = this.hexToRgba(skin.accentColor, 0.4)
        this.ctx.lineWidth = Math.max(1, radius * 0.08)
        this.ctx.beginPath()
        this.ctx.arc(x, y, radius * 0.72, (time / 500) % (Math.PI * 2), (time / 500) % (Math.PI * 2) + Math.PI)
        this.ctx.stroke()
        break
      }
      case 'magma': {
        this.ctx.fillStyle = this.hexToRgba(skin.accentColor, 0.85)
        ;[
          [-0.22, -0.16, 0.19],
          [0.26, 0.06, 0.15],
          [0.02, 0.28, 0.12],
        ].forEach(([dx, dy, scale]) => {
          this.ctx.beginPath()
          this.ctx.arc(x + radius * dx, y + radius * dy, radius * scale, 0, Math.PI * 2)
          this.ctx.fill()
        })
        this.ctx.strokeStyle = this.hexToRgba(skin.secondaryColor, 0.8)
        this.ctx.lineWidth = Math.max(1.5, radius * 0.1)
        this.ctx.beginPath()
        this.ctx.moveTo(x - radius * 0.6, y + radius * 0.05)
        this.ctx.lineTo(x - radius * 0.1, y - radius * 0.22)
        this.ctx.lineTo(x + radius * 0.55, y + radius * 0.16)
        this.ctx.stroke()
        break
      }
      case 'royal': {
        this.ctx.strokeStyle = this.hexToRgba(skin.accentColor, 0.9)
        this.ctx.lineWidth = Math.max(1.2, radius * 0.09)
        this.ctx.beginPath()
        this.ctx.moveTo(x, y - radius * 0.6)
        this.ctx.lineTo(x + radius * 0.35, y)
        this.ctx.lineTo(x, y + radius * 0.52)
        this.ctx.lineTo(x - radius * 0.35, y)
        this.ctx.closePath()
        this.ctx.stroke()
        break
      }
      case 'prism': {
        this.ctx.strokeStyle = this.hexToRgba('#ffffff', 0.7)
        this.ctx.lineWidth = Math.max(1.2, radius * 0.08)
        const sheen = ((time / 450) % 2) - 1
        this.ctx.beginPath()
        this.ctx.moveTo(x + radius * sheen, y - radius)
        this.ctx.lineTo(x + radius * (sheen + 0.35), y + radius)
        this.ctx.stroke()
        break
      }
      case 'neon': {
        this.ctx.strokeStyle = this.hexToRgba('#ffffff', 0.55)
        this.ctx.lineWidth = Math.max(1.3, radius * 0.1)
        this.ctx.beginPath()
        this.ctx.arc(x, y, radius * 0.72, angle - 1.1, angle + 1.2)
        this.ctx.stroke()
        break
      }
      default: {
        this.ctx.fillStyle = this.hexToRgba(skin.accentColor, 0.22)
        this.ctx.beginPath()
        this.ctx.arc(x - radius * 0.18, y - radius * 0.2, radius * 0.55, 0, Math.PI * 2)
        this.ctx.fill()
      }
    }

    this.ctx.restore()
  }

  drawBackground(
    bounds,
    cameraX = bounds.width / 2,
    cameraY = bounds.height / 2,
    viewWidth = window.innerWidth,
    viewHeight = window.innerHeight,
    themeId = 'arcade_maze'
  ) {
    const theme = getThemePreset(themeId)
    const startX = Math.max(0, Math.floor((cameraX - viewWidth / 2) / 140) * 140)
    const endX = Math.min(bounds.width, startX + viewWidth + 280)
    const startY = Math.max(0, Math.floor((cameraY - viewHeight / 2) / 140) * 140)
    const endY = Math.min(bounds.height, startY + viewHeight + 280)

    this.ctx.fillStyle = theme.baseFill
    this.ctx.fillRect(startX, startY, endX - startX, endY - startY)

    if (this.lowEffects) {
      this.ctx.save()
      this.ctx.strokeStyle = theme.accentSoft
      this.ctx.lineWidth = 1
      for (let y = startY + 36; y <= endY; y += 180) {
        this.ctx.beginPath()
        this.ctx.moveTo(startX, y)
        this.ctx.lineTo(endX, y)
        this.ctx.stroke()
      }
      this.ctx.restore()
      return
    }

    this.ctx.save()
    switch (theme.id) {
      case 'retro_lcd': {
        this.ctx.fillStyle = theme.accentSoft
        for (let y = startY; y <= endY; y += 140) {
          for (let x = startX; x <= endX; x += 140) {
            this.ctx.fillRect(x + 16, y + 16, 88, 56)
          }
        }
        break
      }
      case 'arcade_maze': {
        this.ctx.strokeStyle = 'rgba(255, 255, 255, 0.04)'
        this.ctx.lineWidth = 1
        for (let y = startY; y <= endY; y += 44) {
          this.ctx.beginPath()
          this.ctx.moveTo(startX, y)
          this.ctx.lineTo(endX, y)
          this.ctx.stroke()
        }
        this.ctx.fillStyle = theme.accentSoft
        for (let y = startY + 24; y <= endY; y += 160) {
          for (let x = startX + 24; x <= endX; x += 160) {
            this.ctx.beginPath()
            this.ctx.arc(x, y, 6, 0, Math.PI * 2)
            this.ctx.fill()
          }
        }
        break
      }
      case 'pixel_handheld': {
        this.ctx.fillStyle = theme.accentSoft
        for (let y = startY; y <= endY; y += 120) {
          for (let x = startX; x <= endX; x += 120) {
            this.ctx.fillRect(x + 20, y + 20, 28, 28)
            this.ctx.fillRect(x + 68, y + 68, 28, 28)
          }
        }
        break
      }
      case 'brick_phone': {
        this.ctx.strokeStyle = theme.accentSoft
        this.ctx.lineWidth = 2
        for (let x = startX - 200; x <= endX + 200; x += 120) {
          this.ctx.beginPath()
          this.ctx.moveTo(x, startY)
          this.ctx.lineTo(x + 180, endY)
          this.ctx.stroke()
        }
        break
      }
      case 'vector_grid': {
        this.ctx.strokeStyle = theme.accentSoft
        this.ctx.lineWidth = 1
        for (let y = startY; y <= endY; y += 180) {
          for (let x = startX; x <= endX; x += 180) {
            this.ctx.beginPath()
            this.ctx.moveTo(x, y + 30)
            this.ctx.lineTo(x + 30, y)
            this.ctx.lineTo(x + 60, y + 30)
            this.ctx.lineTo(x + 30, y + 60)
            this.ctx.closePath()
            this.ctx.stroke()
          }
        }
        break
      }
      case 'synth_horizon':
      default: {
        this.ctx.strokeStyle = theme.accentSoft
        this.ctx.lineWidth = 2
        for (let y = startY + 30; y <= endY; y += 90) {
          this.ctx.beginPath()
          this.ctx.moveTo(startX, y)
          this.ctx.lineTo(endX, y)
          this.ctx.stroke()
        }
      }
    }
    this.ctx.restore()
  }

  drawGrid(
    bounds,
    cameraX = bounds.width / 2,
    cameraY = bounds.height / 2,
    viewWidth = window.innerWidth,
    viewHeight = window.innerHeight,
    themeId = 'arcade_maze'
  ) {
    const theme = getThemePreset(themeId)
    const step = this.lowEffects ? 180 : 100
    this.ctx.strokeStyle = theme.gridColor;
    this.ctx.lineWidth = this.lowEffects ? 0.6 : 1;
    this.ctx.beginPath();

    // Calculate visible bounds to only draw what's on screen
    const startX = Math.max(0, Math.floor((cameraX - viewWidth / 2) / step) * step);
    const endX = Math.min(bounds.width, startX + viewWidth + step * 2);

    const startY = Math.max(0, Math.floor((cameraY - viewHeight / 2) / step) * step);
    const endY = Math.min(bounds.height, startY + viewHeight + step * 2);

    for (let x = startX; x <= endX; x += step) {
      this.ctx.moveTo(x, startY);
      this.ctx.lineTo(x, endY);
    }
    for (let y = startY; y <= endY; y += step) {
      this.ctx.moveTo(startX, y);
      this.ctx.lineTo(endX, y);
    }
    this.ctx.stroke();
  }

  drawBounds(bounds, themeId = 'arcade_maze') {
    const theme = getThemePreset(themeId)
    this.ctx.strokeStyle = theme.boundsColor;
    this.ctx.lineWidth = 10;
    if (!this.isMobile) {
      this.ctx.shadowBlur = 30;
      this.ctx.shadowColor = theme.boundsColor;
    }
    this.ctx.strokeRect(0, 0, bounds.width, bounds.height);
    this.ctx.shadowBlur = 0;
  }

  drawObstacles(obstacles, themeId = 'arcade_maze') {
    if (!obstacles || obstacles.length === 0) return

    const theme = getThemePreset(themeId)
    this.ctx.save()

    for (const obstacle of obstacles) {
      const x = obstacle.x
      const y = obstacle.y
      const width = obstacle.width
      const height = obstacle.height
      const radius = Math.min(24, Math.min(width, height) * 0.18)

      this.ctx.save()
      this.ctx.fillStyle = this.hexToRgba(theme.boundsColor, 0.16)
      this.ctx.strokeStyle = this.hexToRgba(theme.boundsColor, 0.72)
      this.ctx.lineWidth = 3
      this.roundRect(x, y, width, height, radius)
      this.ctx.fill()
      this.ctx.stroke()

      if (this.lowEffects) {
        this.ctx.restore()
        continue
      }

      switch (obstacle.style) {
        case 'lcd_bar': {
          this.ctx.fillStyle = this.hexToRgba(theme.accentSoft, 0.12)
          for (let scan = y + 8; scan < y + height; scan += 16) {
            this.ctx.fillRect(x + 8, scan, width - 16, 5)
          }
          break
        }
        case 'pixel_block':
        case 'screen_block': {
          this.ctx.fillStyle = this.hexToRgba(theme.gridColor, 0.22)
          for (let row = y + 12; row < y + height - 12; row += 22) {
            for (let col = x + 12; col < x + width - 12; col += 22) {
              this.ctx.fillRect(col, row, 10, 10)
            }
          }
          break
        }
        case 'keypad_key': {
          this.ctx.strokeStyle = this.hexToRgba(theme.accentSoft, 0.32)
          this.ctx.lineWidth = 1.5
          for (let row = y + 14; row < y + height - 6; row += 18) {
            this.ctx.beginPath()
            this.ctx.moveTo(x + 10, row)
            this.ctx.lineTo(x + width - 10, row)
            this.ctx.stroke()
          }
          break
        }
        case 'vector_gate': {
          this.ctx.strokeStyle = this.hexToRgba('#ffffff', 0.22)
          this.ctx.lineWidth = 2
          this.ctx.beginPath()
          this.ctx.moveTo(x + width * 0.15, y + height * 0.15)
          this.ctx.lineTo(x + width * 0.85, y + height * 0.85)
          this.ctx.moveTo(x + width * 0.85, y + height * 0.15)
          this.ctx.lineTo(x + width * 0.15, y + height * 0.85)
          this.ctx.stroke()
          break
        }
        case 'maze_wall':
        default: {
          this.ctx.fillStyle = this.hexToRgba(theme.accentSoft, 0.15)
          for (let col = x + 10; col < x + width - 10; col += 34) {
            this.ctx.fillRect(col, y + 10, 14, height - 20)
          }
        }
      }

      this.ctx.restore()
    }

    this.ctx.restore()
  }

  drawStorm(stormCenter, stormRadius, bounds) {
    if (!stormCenter || !stormRadius) return;

    this.ctx.save();

    // Draw the safe zone
    this.ctx.beginPath();
    this.ctx.arc(stormCenter.x, stormCenter.y, stormRadius, 0, Math.PI * 2);
    this.ctx.lineWidth = 5;
    this.ctx.strokeStyle = '#ff0055';
    this.ctx.shadowBlur = 20;
    this.ctx.shadowColor = '#ff0055';
    this.ctx.stroke();

    // Fill the storm area (everything outside the circle)
    this.ctx.beginPath();
    this.ctx.rect(0, 0, bounds.width, bounds.height); // Outer bounds
    this.ctx.arc(stormCenter.x, stormCenter.y, stormRadius, 0, Math.PI * 2, true); // Inner safe hole (counter-clockwise)
    this.ctx.fillStyle = 'rgba(255, 0, 85, 0.2)';
    this.ctx.fill();

    this.ctx.restore();
  }

  drawBlackHoles(blackHoles, time) {
    if (!blackHoles) return;

    this.ctx.save();
    for (const bh of blackHoles) {
      this.ctx.beginPath();
      this.ctx.arc(bh.x, bh.y, bh.radius, 0, Math.PI * 2);
      this.ctx.fillStyle = '#050510';
      this.ctx.fill();
      this.ctx.strokeStyle = '#3300ff';
      this.ctx.lineWidth = 4;
      if (!this.isMobile) {
        this.ctx.shadowBlur = 30;
        this.ctx.shadowColor = '#3300ff';
      }
      this.ctx.stroke();

      if (!this.isMobile) {
        this.ctx.save();
        this.ctx.translate(bh.x, bh.y);
        this.ctx.rotate(time * 0.002);
        this.ctx.beginPath();
        this.ctx.arc(0, 0, bh.radius * 2, 0, Math.PI);
        this.ctx.strokeStyle = 'rgba(100, 50, 255, 0.5)';
        this.ctx.lineWidth = 2;
        this.ctx.stroke();
        this.ctx.rotate(Math.PI / 2);
        this.ctx.beginPath();
        this.ctx.arc(0, 0, bh.radius * 1.5, 0, Math.PI);
        this.ctx.strokeStyle = 'rgba(50, 0, 255, 0.8)';
        this.ctx.lineWidth = 3;
        this.ctx.stroke();
        this.ctx.restore();
      }

      this.ctx.beginPath();
      this.ctx.arc(bh.x, bh.y, bh.pullRadius, 0, Math.PI * 2);
      this.ctx.fillStyle = 'rgba(20, 0, 50, 0.1)';
      this.ctx.fill();
    }
    this.ctx.restore();
  }

  drawWormholes(wormholes, time) {
    if (!wormholes) return;

    this.ctx.save();
    for (const wh of wormholes) {
      this.ctx.beginPath();
      this.ctx.arc(wh.x, wh.y, wh.radius, 0, Math.PI * 2);
      this.ctx.fillStyle = '#000';
      this.ctx.fill();

      this.ctx.strokeStyle = wh.color;
      this.ctx.lineWidth = 5 + Math.sin(time * 0.005) * 2;
      if (!this.isMobile) {
        this.ctx.shadowBlur = 20 + Math.sin(time * 0.005) * 10;
        this.ctx.shadowColor = wh.color;
      }
      this.ctx.stroke();

      // Draw inner spiral
      this.ctx.save();
      this.ctx.translate(wh.x, wh.y);
      this.ctx.rotate(-time * 0.003); // Opposite rotation of black holes
      this.ctx.beginPath();
      for (let i = 0; i < 5; i++) {
        const angle = (i / 5) * Math.PI * 2;
        this.ctx.moveTo(0, 0);
        this.ctx.quadraticCurveTo(
          Math.cos(angle + 0.5) * wh.radius * 0.8,
          Math.sin(angle + 0.5) * wh.radius * 0.8,
          Math.cos(angle) * wh.radius,
          Math.sin(angle) * wh.radius
        );
      }
      this.ctx.strokeStyle = 'rgba(255,255,255,0.8)';
      this.ctx.lineWidth = 2;
      this.ctx.stroke();
      this.ctx.restore();
    }
    this.ctx.restore();
    this.ctx.globalAlpha = 1.0;
  }

  drawPowerUps(powerUps, time) {
    if (!powerUps) return;
    for (const id in powerUps) {
      const pu = powerUps[id];
      const pulse = Math.sin(time / 200) * 5;
      const radius = pu.radius + pulse;

      // Outer glow
      if (!this.isMobile) {
        this.ctx.shadowBlur = 20;
        this.ctx.shadowColor = pu.color;
      }

      this.ctx.beginPath();
      this.ctx.arc(pu.x, pu.y, radius, 0, Math.PI * 2);
      this.ctx.fillStyle = pu.color;
      this.ctx.fill();

      // Inner icon / detail
      this.ctx.beginPath();
      this.ctx.arc(pu.x, pu.y, radius * 0.6, 0, Math.PI * 2);
      this.ctx.fillStyle = '#fff';
      this.ctx.fill();

      this.ctx.shadowBlur = 0;

      // Label
      if (!this.lowEffects) {
        this.ctx.fillStyle = '#fff';
        this.ctx.font = 'bold 12px Inter';
        this.ctx.textAlign = 'center';
        this.ctx.fillText(pu.type, pu.x, pu.y + radius + 15);
      }
    }
  }

  drawPellets(pellets, camX, camY, viewWidth, viewHeight) {
    const margin = 50;
    const minX = camX - viewWidth / 2 - margin;
    const maxX = camX + viewWidth / 2 + margin;
    const minY = camY - viewHeight / 2 - margin;
    const maxY = camY + viewHeight / 2 + margin;

    for (const id in pellets) {
      const p = pellets[id];
      // Skip pellets outside view
      if (p.x < minX || p.x > maxX || p.y < minY || p.y > maxY) continue;

      this.ctx.beginPath();
      this.ctx.arc(p.x, p.y, p.value * 1.5, 0, Math.PI * 2);
      this.ctx.fillStyle = p.color;
      // Shadow removed for pellets to improve performance
      this.ctx.fill();
    }
  }

  drawPlayers(players, myId, kingId, camX, camY, viewWidth, viewHeight, activeEmotes = {}, time = 0) {
    const visiblePlayers = [];
    const margin = 200; // Extra margin for names/glow
    const minX = camX - viewWidth / 2 - margin;
    const maxX = camX + viewWidth / 2 + margin;
    const minY = camY - viewHeight / 2 - margin;
    const maxY = camY + viewHeight / 2 + margin;

    for (const p of Object.values(players)) {
      if (!p || !p.position) continue;

      // Skip players completely outside view
      const camMargin = 400; // wide margin
      if (
        p.position.x < minX - camMargin ||
        p.position.x > maxX + camMargin ||
        p.position.y < minY - camMargin ||
        p.position.y > maxY + camMargin
      ) {
        continue;
      }

      visiblePlayers.push(p);
    }

    const playerList = this.lowEffects ? visiblePlayers : visiblePlayers.sort((a, b) => a.mass - b.mass)

    for (const p of playerList) {

      const isMe = p.id === myId;
      const isKing = p.id === kingId;
      const playerColor = p.color || '#00ffcc'; // Fallback
      const skin = this.getSkinStyle(p.skin, playerColor)
      const emoteData = activeEmotes[p.id];
      const overlayFrequency = this.lowEffects ? 6 : 3

      this.ctx.lineCap = 'round';
      this.ctx.lineJoin = 'round';

      // Draw body segments
      for (let i = p.segments.length - 1; i >= 0; i--) {
        const seg = p.segments[i];

        // Dynamic tapering size
        const sizeCoef = 1 - (i / Math.max(1, p.segments.length)) * 0.4;
        const radius = p.radius * sizeCoef;

        this.ctx.beginPath();
        this.ctx.arc(seg.x, seg.y, radius, 0, Math.PI * 2);

        // Boost glow effect logic
        if (p.isBoosting || (p.ability.isActive && p.ability.type === 'DASH')) {
          if (!this.isMobile && isMe) {
            this.ctx.shadowBlur = 20;
            this.ctx.shadowColor = '#ffffff';
          }
          this.ctx.fillStyle = '#ffffff';
        } else {
          if (!this.isMobile && isMe) {
            this.ctx.shadowBlur = isMe ? 15 : 5;
            this.ctx.shadowColor = playerColor;
          }
          this.ctx.fillStyle = this.createSkinFill(seg.x, seg.y, radius, skin, time);
        }

        this.ctx.fill();
        this.ctx.strokeStyle = 'rgba(0,0,0,0.3)'; // Slightly more transparent stroke
        this.ctx.lineWidth = 1.5;
        this.ctx.stroke();

        if (!(p.isBoosting || (p.ability.isActive && p.ability.type === 'DASH'))) {
          const shouldDrawPattern =
            !this.lowEffects || i === 0 || i === p.segments.length - 1 || i % overlayFrequency === 0
          if (shouldDrawPattern) {
            this.drawSkinOverlay(seg.x, seg.y, radius, skin, 0, time, i === 0 ? 1 : 0.82)
          }
        }
      }
      this.ctx.shadowBlur = 0;

      // Calculate visual angle for eyes and skin accents
      let angle = 0;
      if (p.velocity.x !== 0 || p.velocity.y !== 0) {
        angle = Math.atan2(p.velocity.y, p.velocity.x);
      }

      // Draw Head
      this.ctx.beginPath();
      const visualRadius = p.radius * (1.05 + (p.pulse || 0) * 0.3);
      this.ctx.arc(p.position.x, p.position.y, visualRadius, 0, Math.PI * 2);
      this.ctx.fillStyle = p.isBoosting ? '#fff' : this.createSkinFill(p.position.x, p.position.y, visualRadius, skin, time, true);

      // Head Glow only for self or king
      if (!this.isMobile && (isMe || isKing)) {
        this.ctx.shadowBlur = 25 + (p.pulse || 0) * 30;
        this.ctx.shadowColor = isMe ? '#ffffff' : playerColor;
      }

      this.ctx.fill();
      this.ctx.shadowBlur = 0;

      if (!p.isBoosting) {
        this.drawSkinOverlay(p.position.x, p.position.y, visualRadius, skin, angle, time, 1.1)
      }

      // Eyes
      const eyeOffset = p.radius * 0.55;
      const eyeSize = p.radius * 0.3;

      // Left eye base
      this.ctx.beginPath();
      let ex1 = p.position.x + Math.cos(angle - 0.6) * eyeOffset;
      let ey1 = p.position.y + Math.sin(angle - 0.6) * eyeOffset;
      this.ctx.arc(ex1, ey1, eyeSize, 0, Math.PI * 2);
      this.ctx.fillStyle = '#0f0f0f';
      this.ctx.fill();
      // Left Iris
      this.ctx.beginPath();
      this.ctx.arc(
        ex1 + Math.cos(angle) * eyeSize * 0.4,
        ey1 + Math.sin(angle) * eyeSize * 0.4,
        eyeSize * 0.4,
        0,
        Math.PI * 2
      );
      this.ctx.fillStyle = '#fff';
      this.ctx.fill();

      // Right eye base
      this.ctx.beginPath();
      let ex2 = p.position.x + Math.cos(angle + 0.6) * eyeOffset;
      let ey2 = p.position.y + Math.sin(angle + 0.6) * eyeOffset;
      this.ctx.arc(ex2, ey2, eyeSize, 0, Math.PI * 2);
      this.ctx.fillStyle = '#0f0f0f';
      this.ctx.fill();
      // Right Iris
      this.ctx.beginPath();
      this.ctx.arc(
        ex2 + Math.cos(angle) * eyeSize * 0.4,
        ey2 + Math.sin(angle) * eyeSize * 0.4,
        eyeSize * 0.4,
        0,
        Math.PI * 2
      );
      this.ctx.fillStyle = '#fff';
      this.ctx.fill();

      // Draw Shield Aura
      if (p.ability.isActive && p.ability.type === 'SHIELD') {
        this.ctx.beginPath();
        this.ctx.arc(p.position.x, p.position.y, p.radius * 2, 0, Math.PI * 2);
        this.ctx.strokeStyle = 'rgba(100, 200, 255, 0.8)';
        this.ctx.lineWidth = this.lowEffects ? 2 : 4;
        this.ctx.stroke();

        if (!this.lowEffects) {
          this.ctx.beginPath();
          this.ctx.arc(p.position.x, p.position.y, p.radius * 2, 0, Math.PI * 2);
          this.ctx.fillStyle = 'rgba(100, 200, 255, 0.2)';
          this.ctx.fill();
        }
      }

      // Draw Magnet Aura
      if (p.ability.isActive && p.ability.type === 'MAGNET') {
        this.ctx.beginPath();
        this.ctx.arc(p.position.x, p.position.y, p.radius + 150, 0, Math.PI * 2);
        this.ctx.strokeStyle = 'rgba(255, 100, 255, 0.3)';
        this.ctx.lineWidth = 2;
        this.ctx.setLineDash([10, 10]);
        this.ctx.stroke();
        this.ctx.setLineDash([]);
      }

      // Draw Crown if King
      if (isKing) {
        this.ctx.save();
        this.ctx.translate(p.position.x, p.position.y);
        this.ctx.rotate(angle + Math.PI / 2); // Sit on top of head
        this.ctx.fillStyle = '#FFD700'; // Gold
        if (!this.isMobile) {
          this.ctx.shadowBlur = 15;
          this.ctx.shadowColor = '#FFD700';
        }

        const crW = p.radius * 1.5;
        const crH = p.radius;
        const offsetY = p.radius * 0.8;

        this.ctx.beginPath();
        this.ctx.moveTo(-crW / 2, -offsetY); // Bottom left
        this.ctx.lineTo(crW / 2, -offsetY); // Bottom right
        this.ctx.lineTo(crW / 2 + crW * 0.2, -offsetY - crH); // Top right tip
        this.ctx.lineTo(crW * 0.2, -offsetY - crH * 0.5); // Inner right
        this.ctx.lineTo(0, -offsetY - crH * 1.2); // Top middle tip
        this.ctx.lineTo(-crW * 0.2, -offsetY - crH * 0.5); // Inner left
        this.ctx.lineTo(-crW / 2 - crW * 0.2, -offsetY - crH); // Top left tip
        this.ctx.fill();
        this.ctx.restore();
      }

      // Mass and Name Tag
      if (isMe) {
        this.ctx.fillStyle = isKing ? '#FFD700' : 'rgba(255, 255, 255, 0.8)';
        this.ctx.font = 'bold 14px Inter';
        this.ctx.textAlign = 'center';
        this.ctx.fillText(
          `${p.name || 'You'} - Mass: ${Math.floor(p.mass)}`,
          p.position.x,
          p.position.y - p.radius * 2 - (isKing ? 15 : 0)
        );
      } else if (p.name && (!this.lowEffects || isKing)) {
        // Show names for bots and other players
        this.ctx.fillStyle = isKing ? '#FFD700' : 'rgba(255, 255, 255, 0.5)';
        this.ctx.font = '12px Inter';
        this.ctx.textAlign = 'center';
        this.ctx.fillText(
          `${p.name}`,
          p.position.x,
          p.position.y - p.radius * 2 - (isKing ? 15 : 0)
        );
      }

      // Draw Emote
      if (emoteData) {
        const emojiMap = {
          COOL: '😎',
          SAD: '😢',
          ANGRY: '😡',
          GG: '🤝',
          LOVE: '❤️',
          LOL: '😂',
        };
        const emoji = emojiMap[emoteData.emoteId] || '❓';

        this.ctx.save();
        this.ctx.font = '30px Arial';
        this.ctx.textAlign = 'center';

        // Floating animation
        const elapsed = Date.now() - emoteData.time;
        const floatY = Math.sin(elapsed / 200) * 10 - 20;
        const opacity = Math.min(1, 2 - elapsed / 1000);

        this.ctx.globalAlpha = opacity;
        this.ctx.fillText(
          emoji,
          p.position.x,
          p.position.y - p.radius * 3 - (isKing ? 30 : 15) + floatY
        );
        this.ctx.restore();
      }
    }
  }

  drawHUD(
    player,
    canvasWidth,
    canvasHeight,
    placement = 1,
    playerCount = 0,
    pelletCount = 0,
    lag = 0,
    renderDelay = 150,
    modeLabel = 'Free-For-All',
    themeLabel = 'Level reactive',
    spawnShieldMs = 0,
    roomState = 'PLAYING',
    killStreak = 0,
    qualityTier = 'balanced'
  ) {
    if (!player) return;

    const ability = player.ability;
    const padding = 30;

    this.ctx.fillStyle = 'white';
    this.ctx.font = 'bold 20px Inter, sans-serif';
    this.ctx.textAlign = 'right';

    // Position HUD at bottom right
    const textX = canvasWidth - padding;
    const textY = canvasHeight - padding - 40;

    this.ctx.fillText(`ABILITY: ${ability.type}`, textX, textY);
    this.ctx.font = '14px Inter, sans-serif';
    this.ctx.fillText(`(Press SPACE)`, textX, textY + 20);

    // Cooldown bar
    const barWidth = 200;
    const barHeight = 20;
    const barX = textX - barWidth;
    const barY = canvasHeight - padding - 20;

    this.ctx.fillStyle = 'rgba(255, 255, 255, 0.2)';
    this.ctx.fillRect(barX, barY, barWidth, barHeight);

    const fillPercent = ability.cooldown > 0 ? 1 - ability.cooldown / ability.maxCooldown : 1;
    this.ctx.fillStyle = fillPercent === 1 ? '#00ff55' : '#ff5500';
    this.ctx.fillRect(barX, barY, barWidth * fillPercent, barHeight);

    if (ability.isActive) {
      this.ctx.fillStyle = '#00ffff';
      this.ctx.font = 'bold 24px Inter, sans-serif';
      this.ctx.fillText('ACTIVE!', textX, textY - 30);
    }

    this.ctx.font = '12px Inter, sans-serif';
    this.ctx.textAlign = 'left';
    this.ctx.fillText('SPACE: Ability | Mouse/Touch: Boost | 1-6: Emotes', 20, canvasHeight - 20);

    const cardWidth = this.lowEffects ? 162 : 180;
    const cardHeight = 64;
    const cardX = 20;
    const cardY = 20;
    this.ctx.fillStyle = 'rgba(6, 10, 18, 0.76)';
    this.roundRect(cardX, cardY, cardWidth, cardHeight, 14);
    this.ctx.fill();
    this.ctx.strokeStyle = 'rgba(255,255,255,0.08)';
    this.ctx.stroke();
    this.ctx.textAlign = 'left';
    this.ctx.fillStyle = '#ffffff';
    this.ctx.font = 'bold 13px Inter, sans-serif';
    this.ctx.fillText(`Rank #${Math.max(1, placement)}/${Math.max(1, playerCount)}`, cardX + 14, cardY + 22);
    this.ctx.fillStyle = 'rgba(226, 232, 240, 0.84)';
    this.ctx.font = '12px Inter, sans-serif';
    this.ctx.fillText(`Mass ${Math.floor(player.mass || 0)} • Pellets ${pelletCount}`, cardX + 14, cardY + 42);
    this.ctx.fillText(roomState === 'PLAYING' ? modeLabel : `${modeLabel} • ${roomState}`, cardX + 14, cardY + 58);

    // Active Power-Ups Central Display
    if (player.activePowerUps && Object.keys(player.activePowerUps).length > 0) {
      let powerUpIdx = 0;
      for (const type in player.activePowerUps) {
        const remaining = player.activePowerUps[type];
        const x = canvasWidth / 2 - 100 + powerUpIdx * 110;
        const y = canvasHeight - 120;

        const colorMap = {
          SHIELD: '#00ffff',
          MAGNET: '#ff00ff',
          FRENZY: '#ff0000',
          GHOST: '#ffffff',
        };
        const color = colorMap[type] || '#fff';

        this.ctx.fillStyle = 'rgba(0,0,0,0.6)';
        this.ctx.fillRect(x, y, 100, 40);
        this.ctx.strokeStyle = color;
        this.ctx.lineWidth = 2;
        this.ctx.strokeRect(x, y, 100, 40);

        this.ctx.fillStyle = color;
        this.ctx.font = 'bold 12px Inter';
        this.ctx.textAlign = 'center';
        this.ctx.fillText(type, x + 50, y + 18);

        const sec = (remaining / 1000).toFixed(1);
        this.ctx.fillStyle = '#fff';
        this.ctx.fillText(`${sec}s`, x + 50, y + 34);

        powerUpIdx++;
      }
    }

    this.ctx.fillStyle = 'rgba(8, 12, 24, 0.76)';
    this.ctx.fillRect(canvasWidth / 2 - 150, 28, 300, 36);
    this.ctx.strokeStyle = 'rgba(255, 255, 255, 0.12)';
    this.ctx.strokeRect(canvasWidth / 2 - 150, 28, 300, 36);
    this.ctx.fillStyle = '#ffffff';
    this.ctx.font = 'bold 13px Inter, sans-serif';
    this.ctx.fillText(`${modeLabel} • ${themeLabel}`, canvasWidth / 2, 51);

    if (killStreak > 1) {
      this.ctx.fillStyle = 'rgba(8, 12, 24, 0.82)';
      this.roundRect(canvasWidth / 2 - 82, 72, 164, 34, 16);
      this.ctx.fill();
      this.ctx.strokeStyle = killStreak >= 4 ? 'rgba(253, 230, 138, 0.5)' : 'rgba(125, 211, 252, 0.36)';
      this.ctx.stroke();
      this.ctx.fillStyle = killStreak >= 4 ? '#fde68a' : '#7dd3fc';
      this.ctx.font = 'bold 14px Inter, sans-serif';
      this.ctx.fillText(`${killStreak}x elimination streak`, canvasWidth / 2, 93);
    }

    const shieldY = killStreak > 1 ? 116 : 76;

    if (spawnShieldMs > 0) {
      const shieldWidth = Math.min(220, canvasWidth * 0.44);
      const shieldX = canvasWidth / 2 - shieldWidth / 2;
      const shieldRatio = Math.min(1, spawnShieldMs / 3200);
      this.ctx.fillStyle = 'rgba(4, 18, 28, 0.72)';
      this.roundRect(shieldX, shieldY, shieldWidth, 22, 11);
      this.ctx.fill();
      this.ctx.fillStyle = '#67e8f9';
      this.roundRect(shieldX, shieldY, shieldWidth * shieldRatio, 22, 11);
      this.ctx.fill();
      this.ctx.fillStyle = '#ecfeff';
      this.ctx.font = 'bold 12px Inter, sans-serif';
      this.ctx.textAlign = 'center';
      this.ctx.fillText(`Spawn shield ${(spawnShieldMs / 1000).toFixed(1)}s`, canvasWidth / 2, shieldY + 15);
    }

    if (!this.lowEffects) {
      this.ctx.fillStyle = 'rgba(148, 163, 184, 0.9)';
      this.ctx.font = '11px monospace';
      this.ctx.textAlign = 'center';
      const netText = `LAG ${~~lag}ms • DELAY ${renderDelay}ms • ${String(qualityTier).toUpperCase()}`;
      this.ctx.fillText(netText, canvasWidth / 2, spawnShieldMs > 0 ? shieldY + 36 : killStreak > 1 ? 122 : 86);
    }
  }

  drawLeaderboard(players, myId) {
    const sortedPlayers = Object.values(players)
      .sort((a, b) => b.score - a.score)
      .slice(0, 5);

    this.ctx.fillStyle = 'rgba(0, 0, 0, 0.5)';
    this.ctx.fillRect(10, 10, 200, 30 + sortedPlayers.length * 25);

    this.ctx.fillStyle = '#ff0055';
    this.ctx.font = 'bold 16px Inter, sans-serif';
    this.ctx.textAlign = 'left';
    this.ctx.fillText('LEADERBOARD', 20, 30);

    this.ctx.font = '14px Inter, sans-serif';
    sortedPlayers.forEach((p, idx) => {
      this.ctx.fillStyle = p.id === myId ? '#00ffcc' : '#ffffff';
      const name = p.name ? p.name : p.id === myId ? 'You' : `Player ${p.id.slice(0, 3)}`;
      this.ctx.fillText(`${idx + 1}. ${name} - ${Math.floor(p.score)}`, 20, 55 + idx * 25);
    });
  }

  drawTeamScores(teamScores, canvasWidth) {
    if (!teamScores) return;

    const maxScore = 50000;
    const barWidth = 400;
    const barHeight = 20;
    const startX = (canvasWidth - barWidth) / 2;
    const startY = 30;

    // Background track
    this.ctx.fillStyle = 'rgba(0,0,0,0.5)';
    this.ctx.fillRect(startX, startY, barWidth, barHeight);
    this.ctx.strokeStyle = '#333';
    this.ctx.strokeRect(startX, startY, barWidth, barHeight);

    // Calculate widths
    const redPercent = Math.min(1, teamScores['RED'] / maxScore);
    const bluePercent = Math.min(1, teamScores['BLUE'] / maxScore);

    // Draw Red
    this.ctx.fillStyle = '#ff0033';
    this.ctx.fillRect(startX, startY, barWidth * redPercent, barHeight);
    // Draw Blue
    this.ctx.fillStyle = '#0033ff';
    // Fill from right to left
    this.ctx.fillRect(
      startX + barWidth - barWidth * bluePercent,
      startY,
      barWidth * bluePercent,
      barHeight
    );

    // Text
    this.ctx.fillStyle = 'white';
    this.ctx.font = 'bold 14px Inter, sans-serif';
    this.ctx.textAlign = 'center';

    if (teamScores['RED'] >= maxScore) {
      this.ctx.fillText('TEAM RED WINS!', canvasWidth / 2, startY + 15);
    } else if (teamScores['BLUE'] >= maxScore) {
      this.ctx.fillText('TEAM BLUE WINS!', canvasWidth / 2, startY + 15);
    } else {
      this.ctx.fillText(
        `${Math.floor(teamScores['RED'])} - GOAL: 50,000 - ${Math.floor(teamScores['BLUE'])}`,
        canvasWidth / 2,
        startY + 15
      );
    }
  }

  drawKillFeed(killFeed, canvasWidth) {
    this.ctx.textAlign = 'right';
    this.ctx.font = 'bold 14px Inter, sans-serif';

    killFeed.forEach((k, idx) => {
      // Fade out based on time
      const age = Date.now() - k.time;
      let alpha = 1;
      if (age > 2000) {
        alpha = 1 - (age - 2000) / 1000;
      }
      alpha = Math.max(0, alpha);

      this.ctx.fillStyle = `rgba(255, 255, 255, ${alpha})`;

      const killerName = k.killerName || `Player ${k.killerId.slice(0, 3)}`;
      const victimName = k.victimName || `Player ${k.victimId.slice(0, 3)}`;

      const text = `${killerName} killed ${victimName}`;
      this.ctx.fillText(text, canvasWidth - 20, 30 + idx * 20);
    });
  }

  drawMinimap(players, myId, bounds, canvasWidth, canvasHeight, cameraX, cameraY, viewportWidth, viewportHeight, kingId) {
    const mmWidth = this.lowEffects ? 118 : 150;
    const mmHeight = mmWidth;
    const mmX = canvasWidth - mmWidth - 20;
    const mmY = canvasHeight - mmHeight - (this.lowEffects ? 92 : 120);

    // Minimap background
    this.ctx.fillStyle = 'rgba(20, 20, 30, 0.7)';
    this.ctx.fillRect(mmX, mmY, mmWidth, mmHeight);
    this.ctx.strokeStyle = '#333';
    this.ctx.lineWidth = 2;
    this.ctx.strokeRect(mmX, mmY, mmWidth, mmHeight);

    // Draw players on minimap
    const scaleX = mmWidth / bounds.width;
    const scaleY = mmHeight / bounds.height;

    for (const id in players) {
      const p = players[id];
      const px = mmX + p.position.x * scaleX;
      const py = mmY + p.position.y * scaleY;

      this.ctx.beginPath();
      const radius = id === myId ? 4 : id === kingId ? 3 : 2;
      this.ctx.arc(px, py, radius, 0, Math.PI * 2);
      this.ctx.fillStyle = id === myId ? '#ffffff' : id === kingId ? '#facc15' : '#ff0055';
      this.ctx.fill();
    }

    const viewRectWidth = Math.max(10, viewportWidth * scaleX);
    const viewRectHeight = Math.max(10, viewportHeight * scaleY);
    const viewRectX = mmX + (cameraX - viewportWidth / 2) * scaleX;
    const viewRectY = mmY + (cameraY - viewportHeight / 2) * scaleY;
    this.ctx.strokeStyle = 'rgba(255,255,255,0.7)';
    this.ctx.lineWidth = 1;
    this.ctx.strokeRect(viewRectX, viewRectY, viewRectWidth, viewRectHeight);

    this.ctx.fillStyle = 'rgba(255,255,255,0.72)';
    this.ctx.font = 'bold 11px Inter, sans-serif';
    this.ctx.textAlign = 'left';
    this.ctx.fillText('ARENA', mmX + 8, mmY + 14);
  }

  drawDeathExplosion(particles) {
    this.ctx.shadowBlur = 0
    particles.forEach((p) => {
      const size = p.size || 2
      const accent = p.secondaryColor || '#ffffff'
      this.ctx.save()
      this.ctx.translate(p.x, p.y)
      this.ctx.globalAlpha = p.life

      switch (p.effect) {
        case 'shockwave':
        case 'supernova':
        case 'nova': {
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size * 1.5, 0, Math.PI * 2)
          this.ctx.fillStyle = p.color
          this.ctx.fill()
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size * 2.6, 0, Math.PI * 2)
          this.ctx.strokeStyle = this.hexToRgba(accent, 0.9)
          this.ctx.lineWidth = Math.max(1, size * 0.35)
          this.ctx.stroke()
          break
        }
        case 'shatter':
        case 'glassbloom': {
          this.ctx.rotate(Math.atan2(p.vy, p.vx))
          this.ctx.beginPath()
          this.ctx.moveTo(-size * 0.6, 0)
          this.ctx.lineTo(0, -size * 1.5)
          this.ctx.lineTo(size * 1.3, size * 0.1)
          this.ctx.lineTo(0, size * 1.1)
          this.ctx.closePath()
          this.ctx.fillStyle = p.color
          this.ctx.fill()
          this.ctx.strokeStyle = this.hexToRgba(accent, 0.9)
          this.ctx.lineWidth = 1
          this.ctx.stroke()
          break
        }
        case 'confetti':
        case 'petal': {
          this.ctx.rotate(Math.atan2(p.vy, p.vx))
          this.ctx.fillStyle = p.color
          this.ctx.fillRect(-size * 0.8, -size * 0.5, size * 1.6, size)
          this.ctx.fillStyle = accent
          this.ctx.fillRect(-size * 0.25, -size * 0.9, size * 0.5, size * 1.8)
          break
        }
        case 'crownflare':
        case 'batburst': {
          this.ctx.rotate(Math.atan2(p.vy, p.vx))
          this.ctx.beginPath()
          this.ctx.moveTo(-size * 1.3, size * 0.7)
          this.ctx.lineTo(-size * 0.2, -size * 0.9)
          this.ctx.lineTo(0, size * 0.1)
          this.ctx.lineTo(size * 0.2, -size * 0.9)
          this.ctx.lineTo(size * 1.3, size * 0.7)
          this.ctx.closePath()
          this.ctx.fillStyle = p.color
          this.ctx.fill()
          break
        }
        case 'warp':
        case 'rift': {
          this.ctx.rotate(Math.atan2(p.vy, p.vx))
          this.ctx.beginPath()
          this.ctx.ellipse(0, 0, size * 1.8, size * 0.85, 0, 0, Math.PI * 2)
          this.ctx.strokeStyle = p.color
          this.ctx.lineWidth = Math.max(1.3, size * 0.3)
          this.ctx.stroke()
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size * 0.5, 0, Math.PI * 2)
          this.ctx.fillStyle = accent
          this.ctx.fill()
          break
        }
        case 'acid':
        case 'magma':
        case 'flare':
        case 'scrapburst':
        default: {
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size, 0, Math.PI * 2)
          this.ctx.fillStyle = p.color
          this.ctx.fill()
          this.ctx.beginPath()
          this.ctx.arc(size * 0.35, -size * 0.28, size * 0.35, 0, Math.PI * 2)
          this.ctx.fillStyle = accent
          this.ctx.fill()
        }
      }

      this.ctx.restore()
    })
    this.ctx.globalAlpha = 1.0
  }

  drawTrailParticles(particles) {
    if (particles.length === 0) return

    particles.forEach((p) => {
      const size = p.size || 2
      const accent = p.secondaryColor || '#ffffff'
      this.ctx.save()
      this.ctx.translate(p.x, p.y)
      this.ctx.globalAlpha = Math.max(0, Math.min(1, p.life))

      switch (p.effect) {
        case 'heroic': {
          this.ctx.rotate(Math.PI / 4)
          this.ctx.fillStyle = p.color
          this.ctx.fillRect(-size, -size, size * 2, size * 2)
          this.ctx.fillStyle = accent
          this.ctx.fillRect(-size * 0.3, -size * 1.5, size * 0.6, size * 3)
          break
        }
        case 'lightslice': {
          this.ctx.rotate(Math.atan2(p.vy, p.vx))
          this.ctx.strokeStyle = p.color
          this.ctx.lineWidth = Math.max(1.4, size * 0.55)
          this.ctx.beginPath()
          this.ctx.moveTo(-size * 1.8, 0)
          this.ctx.lineTo(size * 1.8, 0)
          this.ctx.stroke()
          break
        }
        case 'shadow': {
          this.ctx.fillStyle = this.hexToRgba(p.color, 0.65)
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size * 1.2, 0, Math.PI * 2)
          this.ctx.fill()
          break
        }
        case 'mist':
        case 'aurora': {
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size * 1.35, 0, Math.PI * 2)
          this.ctx.fillStyle = p.color
          this.ctx.fill()
          this.ctx.beginPath()
          this.ctx.arc(size * 0.7, -size * 0.3, size * 0.8, 0, Math.PI * 2)
          this.ctx.fillStyle = accent
          this.ctx.fill()
          break
        }
        case 'sparkle':
        case 'stardust':
        case 'storybook': {
          this.ctx.strokeStyle = p.color
          this.ctx.lineWidth = Math.max(1, size * 0.4)
          this.ctx.beginPath()
          this.ctx.moveTo(-size, 0)
          this.ctx.lineTo(size, 0)
          this.ctx.moveTo(0, -size)
          this.ctx.lineTo(0, size)
          this.ctx.stroke()
          break
        }
        case 'ember':
        case 'sunfire':
        case 'toxic':
        case 'comet':
        default: {
          this.ctx.beginPath()
          this.ctx.arc(0, 0, size, 0, Math.PI * 2)
          this.ctx.fillStyle = p.color
          this.ctx.fill()
        }
      }

      this.ctx.restore()
    })
  }

  drawPowerUpPickup(x, y, color, age) {
    const maxAge = 0.5; // seconds
    const progress = age / maxAge;
    if (progress > 1) return;

    this.ctx.save();
    this.ctx.beginPath();
    // Expanding ring
    const radius = progress * 50;
    this.ctx.arc(x, y, radius, 0, Math.PI * 2);
    this.ctx.strokeStyle = color;
    this.ctx.lineWidth = 4 * (1 - progress);
    this.ctx.globalAlpha = 1 - progress;
    this.ctx.stroke();

    // Secondary pulse
    if (progress > 0.2) {
      this.ctx.beginPath();
      this.ctx.arc(x, y, radius * 0.6, 0, Math.PI * 2);
      this.ctx.stroke();
    }
    this.ctx.restore();
  }

  drawToastStack(toasts, canvasWidth, canvasHeight) {
    if (!toasts || toasts.length === 0) return

    const visibleToasts = toasts.slice(-3)
    visibleToasts.forEach((toast, index) => {
      const age = Date.now() - toast.createdAt
      const fadeIn = Math.min(1, age / 180)
      const fadeOut = Math.min(1, Math.max(0, toast.duration - age) / 320)
      const opacity = Math.max(0, Math.min(fadeIn, fadeOut))
      const y = canvasHeight - 220 - (visibleToasts.length - index - 1) * 72

      this.ctx.save()
      this.ctx.globalAlpha = opacity
      this.ctx.fillStyle = 'rgba(6, 10, 18, 0.86)'
      this.ctx.fillRect(24, y, 320, 56)
      this.ctx.fillStyle = toast.accent || '#7dd3fc'
      this.ctx.fillRect(24, y, 5, 56)
      this.ctx.strokeStyle = 'rgba(255,255,255,0.1)'
      this.ctx.strokeRect(24, y, 320, 56)

      this.ctx.textAlign = 'left'
      this.ctx.fillStyle = '#ffffff'
      this.ctx.font = 'bold 13px Inter, sans-serif'
      this.ctx.fillText(toast.title, 40, y + 20)
      this.ctx.fillStyle = 'rgba(226,232,240,0.84)'
      this.ctx.font = '12px Inter, sans-serif'
      this.ctx.fillText(toast.detail, 40, y + 40)
      this.ctx.restore()
    })
  }

  drawMobileUI(joystick, boostBtn, abilityBtn) {
    // Draw Joystick Base
    if (joystick.active) {
      this.ctx.beginPath();
      this.ctx.arc(joystick.baseX, joystick.baseY, joystick.radius, 0, Math.PI * 2);
      this.ctx.fillStyle = 'rgba(255, 255, 255, 0.12)';
      this.ctx.fill();
      this.ctx.strokeStyle = 'rgba(255, 255, 255, 0.32)';
      this.ctx.lineWidth = 2.5;
      this.ctx.stroke();

      // Draw Stick
      this.ctx.beginPath();
      this.ctx.arc(joystick.stickX, joystick.stickY, joystick.radius * 0.4, 0, Math.PI * 2);
      this.ctx.fillStyle = 'rgba(255, 255, 255, 0.62)';
      this.ctx.fill();
    }

    // Draw Buttons
    const drawBtn = (btn, text, color) => {
      this.ctx.beginPath();
      this.ctx.arc(btn.x, btn.y, btn.radius * 1.22, 0, Math.PI * 2);
      this.ctx.fillStyle = `rgba(${color}, 0.08)`;
      this.ctx.fill();

      this.ctx.beginPath();
      this.ctx.arc(btn.x, btn.y, btn.radius, 0, Math.PI * 2);
      this.ctx.fillStyle = btn.active ? `rgba(${color}, 0.6)` : `rgba(${color}, 0.2)`;
      this.ctx.fill();
      this.ctx.strokeStyle = `rgba(${color}, 0.5)`;
      this.ctx.lineWidth = 2.5;
      this.ctx.stroke();

      this.ctx.fillStyle = 'white';
      this.ctx.font = 'bold 13px Inter, sans-serif';
      this.ctx.textAlign = 'center';
      this.ctx.fillText(text, btn.x, btn.y + 4);
    };

    drawBtn(boostBtn, 'BOOST', '255, 80, 80');
    drawBtn(abilityBtn, 'SKILL', '80, 200, 255');
  }
}
